packages/sanity/src/core/studio/ViteDevServerStopped.tsx

```diff
@@ -0,0 +1,69 @@
+/* eslint-disable i18next/no-literal-string -- will not support i18n in error boundaries */
+import {Card, Container, Heading, Stack, Text} from '@sanity/ui'
+import {type ReactNode, useCallback, useEffect, useState} from 'react'
+import {type ViteHotContext} from 'vite/types/hot.js'
+
+const ERROR_TITLE = 'Dev server stopped'
+const ERROR_DESCRIPTION =
+  'The development server has stopped. You may need to restart it to continue working.'
+
+class ViteDevServerStoppedError extends Error {
+  ViteDevServerStoppedError: boolean
+
+  constructor() {
+    super(ERROR_TITLE)
+    this.name = 'ViteDevServerStoppedError'
+    this.ViteDevServerStoppedError = true
+  }
+}
+const serverHot = import.meta.hot
+const isViteServer = (hot: unknown): hot is ViteHotContext => Boolean(hot)
+
+const useDetectViteDevServerStopped = () => {
+  const [devServerStopped, setDevServerStopped] = useState(false)
+
+  const markDevServerStopped = useCallback(() => setDevServerStopped(true), [])
+
+  useEffect(() => {
+    // no early return to optimize tree-shaking
+    if (isViteServer(serverHot)) {
+      serverHot.on('vite:ws:disconnect', markDevServerStopped)
```

Comment:
@juice49 using `import.meta.hot.on`. In our version of vite (4.x) there is no `.off`, however there is in 5.x. Don't think it's a big issue here, since restarting the server will trigger a refresh on the client. But does raise a question around us upgrading perhaps... one for another time

Reply:
I'm not concerned around multiple subscriptions since starting the server again causing a remounting of the entire app, clearing all existing listeners